clients/PortManager/DaqPortManager.8
.TH DaqPortManager 8 "January 2005" nscldaq-8.0 "ADMINISTRATIVE COMMANDS"

.SH NAME

/usr/opt/daq/current/bin/DaqPortManager \- Persistent server to manage TCP/IP
server listen ports.

.SH SYNOPSIS

tclsh8.4 DaqPortManager [-ports range] [-listen port] [-log file] [-usagefile file] &


.SH DESCRIPTION
.PP
DaqPortManager is a persistent server that should be run at system startup
time.  The program manages a block of TCP/IP ports.   Clients connect to the
server to request ports as well as to ask the server to return the current port
assignments. 
.PP
Command options (see OPTIONS) describe how th eprogram starts up.  The command
options allow the user to define the block of ports that will be managed, to
determine which port the port manager itself will listen on for connections and
to define where the software will writ its log file.  Several other files are
maintained by the program (see FILES).

.SH OPTIONS
.TP
\-ports \fIrange\fR
The \-ports switch is followed by a port range in the form of a pair of numbers
separated by a dash (\-).  The pair of numbers defines a contiguous range of
TCP/IP ports that will be managed by the server.   If this switch and its
parameter are omitted, the server, by default, will manage the range 30001-31000

.TP
\-listen \fIport\fR
The \-listen switch is followed by a single numeric parameter that defines the
port on which the server will listen for client connections.  If this is not
supplied, the server will listen for connections on port 30000.
.TP
\-log \fIfile\fR
The \-log switch is followed by a path to a file that will be used to log the
server's actions.  The server logs connections, port allocations, port
allocation failures, port releases and
illegal requests by clients.  The default log file is
/var/log/nscldaq/portmanager.log 
.TP
\-usagefile \fIfile\fR
The \-usagefile switch is followed by a path to a file that will be used to
hold the instantaneous port usage.  This file can be examined to show the
current port usage. Each line in this file is a three element TCL list of port,
application and user for an allocated port.   If the \-usagefile switch is not
provided, the server defaults to /var/tmp/daqportmgr/ports.txt

.SH PROTOCOL
The server protocol is quite simple. Once connected, two transactions are
recognized.  Each transaction is requested by the client and fulfilled by the
server.   The requests are:
.TP
GIMME \fIappname user\fR
GIMME requests the server allocate a port.  When allocated, the server will
believe that the port is allocated to the application named \fIappname\fR run by
the user \fIuser\fR. If successful, the server replies:
.TP
OK \fIportnum\fR
Where the OK indicates success and the \fIportnum\fR is the number of the port
allocated.   If the server cannot allocate a port it will reply:
.TP
FAIL \fIreason\fR
where \fIreason\fR is a TCL quoted item describing why the port could not be
allocated.  
.PP
Once a client has allocated a port it must hold a persistent connection to the
server.  When a client owning ports drops this connection its ports are freed
by the server. While possibly inconvenient, this protocol prevents clients from
accidently holding on to ports past exit.
.TP
LIST
Requests that the server list port usage.  The server will issue a multiline
reply to this request.  The first line will be of the form:
.TP
OK \fIn\fR
Where \fIn\fR is the number of ports that are currently allocated.   The \fIn\fR
lines that follow each consists of a 3 element TCL list containing in order,
the port number, the application name and the name of the user that is running
the application.  Note that if there are multiple instances of the same
application run by the same user,
They will be qualified by appending an underscore and an application instance number.

.SH EXAMPLES
.nf
tclsh8.4 /usr/opt/daq/current/bin/DaqPortManager -ports 31000-40000 &
.fi

Starts the port manager listening on port 30000 managing ports in the range
31000 through 40000.

.SH FILES
The server maintains:

.nf
/var/run/nscldaq/daqportmgr.pid  - contains the port manager's process id.
/var/log/nscldaq/portmanager.log - Default logfile
/var/tmp/daqportmgr/listen.port  - Contains the listen port for the server.
/var/tmp/daqportmgr/ports.txt    - Default port usage file.
.fi

.SH BUGS
.PP
The server should be able to default to listen on the service daqportmgr in
/etc/services if present before falling back on port 30000
.PP
The 'uniquification' of the application name for the LIST command may vary from
query to query as applications drop out.
.PP
Only connections to/from "localhost" are considered local.  The server does not
bother to determine if any other sources are local even though they may be.
.PP
If multiple instances of the port manager are run, only last one started will be
listed in /var/run/nscldaq/daqportmgr.pid

.SH "SEE ALSO"

CPortManager(3), CPortManagerException(3), portAllocator(3tcl)

ChangeLog
Dec 16, 2010 
      - Fix error in CRingItem copy construction ... if there was a
	payload, the m_pCursor member was not being set to point to the
	end of that payload...leading the CRingItem target to think the
	item was empty.  This was also a problem with operator= which also
	used CRingItem::copyin where the problem lay.
      - Fix various copy in issues with CRingBuffer and subclasses.
	make all the tests pass that should pass and remove those that
	should no longer pass due to changes in the way sampling
	has been done.
Dec 20,2010
      - Fix error in URL parsing that made absolute file urls fail.
	(removed the leading /).
      - Fix error in Gui that prevented event file segments from
	showing up on the status strip.
Dec 31, 2010
      - In readout gui, make links between files symbolic so that they
	port across backup and restores in a different directory tree as
	happens during NSCL experiment closeout.
January 27, 2011
	- Fix up Experiment.tcl so things work on etch systems (for now).
	- Fix errors in copy construction in daq/format.
February 17, 2011	
	- Fix error in event file overwite handling in readoutgui:
	  the softlink must be deleted before being made.
	
March 10, 2011 10.0-004
	- Make ring master wait until the port manager is actually alive
	rather than dying if it isn't.
	- Install VERSION file in make install target.
	- Misc. changes to Makefile.am's and configure.ac so that
	autoreconf isn't so bitchy.
April 8 2011 10.0-005
	- Stop elapsed run timer in readout gui when the run ends.
April 13, 2011 10.0-006
	- Really get the elapsed run timer logic right ...with
	interpolation from the scaler buffers.
	
June 10, 2011 10.0-007
	- Add -textheight switch Readout shell to support running
	  in on displays with limited heights.
	
June 22, 2011 10.0-008
	- Add --deleteonexit switch to stdintoring
	- CRemoteAccess - modify to start stdintoring with --deleteonexit
	  so that the proxy ring is deleted if the remote part of the
	  hoisting pipeline dies.
July 8, 2011 10.0-009
	- Extend length of getpwbyuid buffer in CReadoutMain.cpp it was
	only marginally good for some people with small gecos fields.
August 5, 2011 10.0-010
	- Remove extraneous -lpub from daqstart's Makefile.am.
October 12, 2011 10.0-011
	Fixes for defects in readout framework found when getting CAESAR ready:
	- SBS Makefile did not honor USERCXXFLAGS, USERCCFLAGS nor
 	  USERLDFLAGS support was added in a way that ensures that
	  USERCCFLAGS and USERCXXFLAGS may differ but nonetheless
	  factoring out the common flagsets.
	- Add -D's  to the Makefiel includes so that users accessing VME
	  classes get the fact that we have a mapped interface.
	- Add missing #define __CREADOUTMAIN_H to CReadoutMain.h
	- Fix error in CAENV262Busy - defined GoReady rather than GoClear
	  so it became an abstract class.  Fixed while retaining the
	  GoReady function because we know the workaround kludge/fix won't
	  disappear.
	- Call clear on main scaler bank in CExperiment after the readout
	  so scaler don't continuously increment
	* Work around issues with the documentation getting too big for 
	  docbook2dvi (it segfaults)...by generating man pages and 
	  htmldocs only.
        - Turn off driver tracing in CVMEInterface::Read. oops
	- Lock/unlock VME Interface in trigger main loop for some dwell
	  time.
	- Fix broken -p switch in tcl server.
	- Clear scalers and readout hardware much later in the
 	  initialization game.
	- Fix time handling for scaler reads.
October 17, 2011 10.0-012
	- Move begin run emit to just prior to starting the 
	  atual run itself...so we're not timing the hardware
	  initialization.

October 18, 2011 10.0-013	
	- Stop using run state time for anything but keep computing each
	from 'now' and start stamp.
October 18, 2011 10.0-014
	- Use ns resolutino realtime clock for all CExperiment timing
	to try to get better synch with the scaler trigger clock.
	- Fix error in CRingSelectionPredicate...would not allow the last
	ring item to be fetched until more were in the ring.
October 18, 2011 10.0-015
	- Pull in up to date scaler display program from 8.2..which
	diverged after this forked off.
October 21, 2011 10.0-016
	- Fix Packet/Vpacket macros so that they generate compatibility
	  (word size-field) packets not checking for jumbo-bufferness.
October 24, 2011 10.0-017
	- Add memory barriers on shm operations that write to ensure
	everthing is flushed to shm. - gcc specific code.
	- Add a bit of a safety zone between get/put pointers.
	- Go to a two phase client allocation: first set pid - 0
	  which makes it ineligible for size computation, and claims it
	  for use.  Then set the get pointer and finally set the PID to
	  the actual value making it elligible for size computation.
Nobvember 4, 2011 10.0-018
	- Deal a bit better with pointer skew that can happen as new
	  consumers enter the fray.
	- Play with the tkdumper help file...fix missing link.
	- Provide migration tutorial.
	
November 4, 2011 10.0-019
	- Make stdintoring understand that data are structured not a
	stream.
November 7, 2011 10.0-020
	- If a proxy ring has no producer then restart the pipeline if a
	client attaches.  This provides a way to recover from remote
	system reboot that's much saner.
        - If a ring is deleted then kill its clients.
	- If a get/put/peek operations discoveres the client's PID has
	been co-opted by some other process or released, throw a state
	exception and refuse to perform the operation.  This deals with an
	edge case that should not happen...however better safe than sorry
	especially since the cost is negligible.
	- Require stdintoring have at least a full header in putData
	else the remainder is a residual.
	- Use same run time offset algorithm as scalers for variable dump
	bufs.
November 9-16, 2011 10.0-021
	- Add --exitonend option to ringselector (WO W00957).
	- Add compatibilitybuffer - ring to daq8.2 buffer utility (W00957)

nextgen/servers/portmanager/DaqPortManager.tcl
#!/bin/sh
# start tclsh \
    exec tclsh ${0} ${@}


#    This software is Copyright by the Board of Trustees of Michigan
#    State University (c) Copyright 2005.
#
#    You may use this software under the terms of the GNU public license
#    (GPL).  The terms of this license are described at:
#
#     http://www.gnu.org/licenses/gpl.txt
#
#    Author:
#             Ron Fox
#	     NSCL
#	     Michigan State University
#	     East Lansing, MI 48824-1321



#
#   This is the application itself.  We are just going to setup the
#   pieces and parts of the system, hook them together and enter the event loop.
#
#   Since we have no idea which version of tcl will be used to run us, we require that
#   the user run us as tclshx.y Application.tcl ...
#

#   Usage:
#      tclshx.y TclServer.tcl [-ports range] [-listen port] [-log file]
#
#  If being run from root, in general the defaults are correct for all of these:
#      -ports  30001-31000
#      -listen 30000
#      -log    /var/log/var/log/nscldaq/portmanager.log
#
 
# In order to find the packages we need to add the canonicalized .. to 
# auto_path if it isn't already.

set here [file dirname [info script]]
set wd [pwd]


cd [file join $here .. TclLibs]
set libDir [pwd]
cd $wd

if {[lsearch -exact $auto_path $libDir] == -1} {
    set auto_path [concat $libDir $auto_path]
}

package require snit
package require ConnectionManager
package require PortManager
package require Log

#  Global variables that hold the defaults:

set defaultListenPort   30000
set defaultPortRange   {30001 31000}
set defaultLogFile      /var/log/nscldaq/portmanager.log
set defaultUsageFile    /var/tmp/daqportmgr/ports.txt
set defaultPortFile     /var/tmp/daqportmgr/listen.port
set defaultPidFile      /var/run/nscldaq/daqportmgr.pid

#  These global variables hold the selected run parameter values:

set selectedListenPort  $defaultListenPort
set selectedPortRange   $defaultPortRange
set selectedLogFile     $defaultLogFile
set selectedUsageFile   $defaultUsageFile
set selectedPortFile    $defaultPortFile
set selectedPidFile     $defaultPidFile

# usage:
#    Produces a return value that contains the program
#    usage string.
#
proc usage {} {
    set self [file tail [info script]]
    set Usage         "$self ?-ports low-high? ?-listen port? ?-log filename?"
    append Usage "\n" {    -ports   - Specifies the range of ports to be managed}
    append Usage "\n" {    -listen  - Specifies the port on which the program listens for connections}
    append Usage "\n" {    -log     - Specifies the log file}
    
    return $Usage
}

# ParseParameters:
#   Parse the command line parameters.  Since we don't support  anything
#   other than switches with parameters this is relatively easy.
# Parameters:
#   rawParameters   (list)
#       The list of raw command line parameters.
# SideEffects:
#   The global variables selected* get set according to the mix of default
#   values and switched values.
#
proc ParseParameters {rawParameters} {
    global  selectedListenPort
    global  selectedPortRange
    global  selectedLogFile
    global  selectedPidFile
    global  selectedPortFile
    
    # Parameters come in pairs:
    
    for {set i 0} {$i < [llength $rawParameters]} {incr i 2} {
        set Switch [lindex $rawParameters $i]
        set Value  [lindex $rawParameters [expr {$i+1}]]
        switch -exact -- $Switch {
            -ports {
                #  Ports must have a list with 2 numeric elements...
                
                if {[scan $Value "%d-%d" low high] != 2} {
                    error "Failed to parse -ports $Value  \n [usage]"
                }
                #  If necessary swap values to get low hi pairs:
                
                if {$low > $high} {
                    set tmp $high
                    set high $low
                    set low  $tmp
                }
                set selectedPortRange [list $low $high]
                
            }
            -listen {
                # must have a single numerical port value:
                
                if {[scan $Value "%d" port] != 1} {
                    error "Failed to parse -listen $Value \n [usage]"
                }
                set selectedListenPort $port
            }
            -log {
                # The value is a text string:
                
                set selectedLogFile $Value
            }
	    -portfile {			# Undocumented
		set selectedPortFile $Value
	    }
	    -pidfile {			# Undocumented
		set selectedPidFile $Value
	    }
            default {
                error "Invalid switch $Switch\n [usage]"
            }
        }
       
    }
    
}
# enterEventLoop
#   Enters the event loop:  This  program is built to run
#   from tclsh, rather than wish, but is event driven.
#   The simple code below makes tclsh process events
#   until the application is forced to exit:
#
set forever 0
proc enterEventLoop {} {
    global forever
    vwait  forever
}
# run:
#   Called to run the application.  Any initialization is done first,
#   then this proc is called; to instantiate instances of the components
#   and to stitch them together.
# Parameters:
#   argList (list)
#       Program parameter list e.g. from command line.
#
proc run {argList} {
    global selectedListenPort
    global selectedPortRange
    global selectedLogFile
    global selectedUsageFile
    global selectedPortFile
    global selectedPidFile
    
    if {[catch {ParseParameters $argList} msg]} {
	puts stderr $msg
	exit -1
    }
    
    # Create the directories needed by all the files:
    
    file mkdir [file dirname $selectedLogFile]
    file mkdir [file dirname $selectedUsageFile]
    file mkdir [file dirname $selectedPortFile]
    file mkdir [file dirname $selectedPidFile]
    
    # Set our port and pid files:
    
    set   fd [open $selectedPortFile w]
    puts $fd $selectedListenPort
    close $fd
    
    set   fd [open $selectedPidFile w]
    puts $fd [pid]
    close $fd
    
    #  Build the Log and port manager objects:
    
    puts "log: $selectedLogFile"
    
    set logger [Log create log -filename $selectedLogFile]
    set ports  [PortManager create ports -range $selectedPortRange]
    set cm     [ConnectionManager create cm -ports     $ports \
                                            -logger    $logger \
                                            -listen    $selectedListenPort \
                                            -usagefile $selectedUsageFile]
                                            
    # Start processing events:
    
    enterEventLoop
}

#  Entry point
if {[info exists argv]} {
    run $argv
}
# 